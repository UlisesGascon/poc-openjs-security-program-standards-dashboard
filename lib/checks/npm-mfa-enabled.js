import { getConfig } from '../../config/index.js'
import { isDateWithinPolicy } from '../utils/index.js'

const { generalExpirationPolicy } = getConfig()

export function checkNpmMfaEnabled (projects, checks) {
  checks.npmMfaEnabled = {
    name: 'MFA enabled for npm',
    description: 'Multi Factor Authentication (MFA) Enforced Across the npm Organization',
    level: {
      incubating: 'expected',
      graduated: 'expected',
      retiring: 'expected'
    },
    url: 'https://openjs-security-program-standards.netlify.app/details/item-3',
    projects: {},
    category: 'user authentication'
  }

  for (const project of projects) {
    checks.npmMfaEnabled.projects[project.name] = {}
    const mfaDetails = project.additionalInformation?.npmMfaEnabled || {}

    const criteria = mfaDetails.globalAchievement &&
            isDateWithinPolicy(mfaDetails.confirmationDate, generalExpirationPolicy)

    if (criteria) {
      checks.npmMfaEnabled.projects[project.name].globalAchievement = true
      checks.npmMfaEnabled.projects[project.name].rationale = mfaDetails.rationale
    } else {
      checks.npmMfaEnabled.projects[project.name].globalAchievement = false
      checks.npmMfaEnabled.projects[project.name].rationale = 'The team has not completed the configuration or requires a confirmation.'
    }
  }
}
