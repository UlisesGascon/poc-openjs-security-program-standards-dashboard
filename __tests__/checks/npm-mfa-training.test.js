import { checkNpmMfaEnabled } from '../../lib/checks/npm-mfa-enabled.js'
import { isDateWithinPolicy } from '../../lib/utils/index.js'

jest.mock('../../lib/utils/index.js')

describe('checkNpmMfaEnabled', () => {
  let projects
  let checks
  let tasks
  let alerts

  beforeEach(() => {
    projects = [
      {
        name: 'Project1',
        additionalInformation: {
          npmMfaEnabled: {
            globalAchievement: true,
            rationale: 'Setup completed',
            confirmationDate: new Date().toISOString()
          }
        }
      },
      {
        name: 'Project2',
        additionalInformation: {
          npmMfaEnabled: {
            globalAchievement: false,
            rationale: 'Setup not completed',
            confirmationDate: new Date().toISOString()
          }
        }
      }
    ]
    checks = {}
    tasks = {}
    alerts = {}
    isDateWithinPolicy.mockReturnValue(true)
  })

  it('should update checks with project compliance data', () => {
    checkNpmMfaEnabled({ projects, checks, tasks, alerts })
    expect(checks).toMatchSnapshot()
    expect(tasks).toMatchSnapshot()
    expect(alerts).toMatchSnapshot()
  })

  it('should handle projects with no additional information', () => {
    projects = [
      {
        name: 'Project3'
      }
    ]
    checkNpmMfaEnabled({ projects, checks, tasks, alerts })
    expect(checks).toMatchSnapshot()
    expect(tasks).toMatchSnapshot()
    expect(alerts).toMatchSnapshot()
  })

  it('should handle projects with expired setup', () => {
    projects = [
      {
        name: 'Project4',
        additionalInformation: {
          npmMfaEnabled: {
            globalAchievement: true,
            rationale: 'Setup completed',
            confirmationDate: new Date().toISOString()
          }
        }
      }
    ]
    isDateWithinPolicy.mockReturnValue(false)
    checkNpmMfaEnabled({ projects, checks, tasks, alerts })
    expect(checks).toMatchSnapshot()
    expect(tasks).toMatchSnapshot()
    expect(alerts).toMatchSnapshot()
  })
})
