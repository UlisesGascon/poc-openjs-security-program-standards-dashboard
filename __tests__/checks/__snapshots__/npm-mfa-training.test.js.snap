// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkNpmMfaEnabled should handle projects with expired setup 1`] = `
{
  "npmMfaEnabled": {
    "category": "user authentication",
    "description": "Multi Factor Authentication (MFA) Enforced Across the npm Organization",
    "level": {
      "graduated": "expected",
      "incubating": "expected",
      "retiring": "expected",
    },
    "name": "MFA enabled for npm",
    "projects": {
      "Project4": {
        "globalAchievement": false,
        "rationale": "The team has not completed the configuration or requires a confirmation.",
      },
    },
    "url": "https://openjs-security-program-standards.netlify.app/details/item-3",
  },
}
`;

exports[`checkNpmMfaEnabled should handle projects with no additional information 1`] = `
{
  "npmMfaEnabled": {
    "category": "user authentication",
    "description": "Multi Factor Authentication (MFA) Enforced Across the npm Organization",
    "level": {
      "graduated": "expected",
      "incubating": "expected",
      "retiring": "expected",
    },
    "name": "MFA enabled for npm",
    "projects": {
      "Project3": {
        "globalAchievement": false,
        "rationale": "The team has not completed the configuration or requires a confirmation.",
      },
    },
    "url": "https://openjs-security-program-standards.netlify.app/details/item-3",
  },
}
`;

exports[`checkNpmMfaEnabled should update checks with project compliance data 1`] = `
{
  "npmMfaEnabled": {
    "category": "user authentication",
    "description": "Multi Factor Authentication (MFA) Enforced Across the npm Organization",
    "level": {
      "graduated": "expected",
      "incubating": "expected",
      "retiring": "expected",
    },
    "name": "MFA enabled for npm",
    "projects": {
      "Project1": {
        "globalAchievement": true,
        "rationale": "Setup completed",
      },
      "Project2": {
        "globalAchievement": false,
        "rationale": "The team has not completed the configuration or requires a confirmation.",
      },
    },
    "url": "https://openjs-security-program-standards.netlify.app/details/item-3",
  },
}
`;
